### PR TITLE
ListOption: adds a new SendInitialEvents option

### PR DESCRIPTION
**Pull request by p0lyn0mial**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
adds a `SendInitialEvents` option that does nothing but simply rejects the request. 
Eventually, it will be used by https://github.com/kubernetes/kubernetes/pull/110960.
In case the above PR doesn't make 1.25, it gives us a clean PR to backport to supported versions to ease the transition.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Create a new `SendInitialEvents` ListOption. If the parameter is set, then the request will be rejected, as the feature is not implemented yet. This will allow backward compatibility.

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3157-watch-list
```

### CODE REVIEW

**Comment by wojtek-t** [comment]
This LGTM (the actual implementation is in #110960)
/lgtm

However, as an API change this requires API approval.
/assign @deads2k @lavalamp

**Comment by p0lyn0mial** [comment]
It looks like`TestCacheControl` tests fail on `server.TearDownFn()` which waits for the server to stop. Are there any known issues with stopping the server?

**Comment by liggitt** [comment]
looks like https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/3157-watch-list/kep.yaml is still provisional, so not queueing this up for api review yet

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go [hunk 1]
```diff
@@ -66,6 +66,24 @@ type ListOptions struct {
...
+	// This option is meant to be used with the resourceVersion parameter with the following semantics:
+	//
+	// - sendInitialEvent and RV>0 - provides the initial events and when we reach the desired RV,
+	//   the synthetic bookmark event. Note that the desired RV must be <= the current revision the server knows about.
+	//   If the desired RV is newer than the current revision the server is aware of,
+	//   The TooLargeResourceVersionError is returned.
+	//   Essentially this mode provides NotOlderThan semantics.
+	// - sendInitialEvent and RV=0 - provides the initial events and the synthetic bookmark with the current RV.
+	// - sendInitialEvent and RV="" - matches data at the most recent ResourceVersion.
+	//   The returned data is consistent, that is, served from etcd via a quorum read.
```
Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:83]: Here's a different idea: make it not depend on RV at all and always work as described here.

I don't think users who know a specific RV would want to use this option anyway.

@jpbetz can you comment on this API?

Comment (deads2k) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:83]: for convenience a link to current RV semantics for watch: https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-watch

Comment (deads2k) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:83]: > I don't think users who know a specific RV would want to use this option anyway.

List currently allows NotOlderThan.  If we do not provide the ability to set `sendInitialEvents` and rv>0, we will make it impossible for some types of lists to switch.  We've already given up on Exact, but also giving up on NotOlderThan seems unnecessarily limiting.

Comment (jpbetz) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:83]: I'm in agreement with you @lavalamp.

RV became overexposed on the LIST operation and IMHO has caused us years of trouble.  The main effect I've observed from having RV exposed is that it makes API callers aware of the watch cache and give them the ability to request data from cache for the performance/scalability benefits, which has resulted in a leaky abstraction.  It's worrying that if you ask people what RV=0 does, they say "it serves the request from cache" instead of saying "it indicates that I can tolerate arbitrarily old data" (which is very rarely the case).

We still have https://github.com/kubernetes/kubernetes/issues/59848 open, even though we patched up the RV semantics to make it possible make a consistent read LIST request in all possible configurations, because of scale concerns with bypassing the watch cache which go back to someone setting RV="" not because it's the correct thing to do (in terms of correctness) but because they wanted the performance/scale boost.


I'd strongly prefer we don't expose RV.  And for watch in particular, which is eventually consistent, it's going to be even more difficult to explain the semantics here.  But maybe say that the 1st bookmark demarks the end of the equivalent of a consistent read LIST response?  Then if we want to allow the watch cache to serve the inital list use something like the [consistent reads from cache proposal](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2340-Consistent-reads-from-cache)?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go [hunk 2]
```diff
@@ -42,5 +42,22 @@ func ValidateListOptions(options *internalversion.ListOptions) field.ErrorList {
...
+	if options.SendInitialEvents {
+		allErrs = append(allErrs, field.Forbidden(field.NewPath("sendInitialEvents"), "sendInitialEvents is forbidden for list"))
+	}
+	return allErrs
+}
+
+func validateWatchOptions(options *internalversion.ListOptions) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if !options.Watch {
+		return allErrs
```
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/validation/validation.go:54]: seems error worthy. Something went wrong.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go (hunk 1 above)
Comment (wojtek-t) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:83]: > if the user does both a list and a watch, they have an exact RV from the list and therefore do not need the watch to sendInitialEvents because they already got them from the list.

Not necessarily.
Currently, in the reflector (which automatically translates to informer and all our client frameworks), when the previous watch finishes with "resource version to old" (or other error that requires relist), what we do is basically we send "List with NotOlderThan semantic with the RV of the last event that we received". This part is already correct (ensures not going back in time *[1]). So given we want to switch from the current "list + continue with watch" to the new "watch with initial events", we know exactly the RV that we want to use here - it's the RV of the last received event.
I don't think there is anything bad in exposing this semantic.

All problems that we have is due to the concept of "RV=0" (because it basically doesn't give any guarantees about data freshness). 
With the above in mind, thinking more about it, maybe all we need is to actually don't expose "RV=0" semantic with the new `SendInitialEvents` feature.
So basically switch to:
- `RV=''"` -> it's boils down to consistent list [and we will migrate reflector to use this semantics on component startup]
- `RV>0` -> the semantic remains as is described - it allows us to basically switch in-place in reflector framework without causing any correctness issue
- `RV=0` - we doesn't allow this call and reject those requests via validation

[And eventually completely deprecate the RV=0 semantic also if `SendInitialEvents` option is not used.]

@deads2k @lavalamp @jpbetz - thoughts?


---

*[1] It guarantees correctness within a single incarnation of a components. We still can go back in time across components restarts.

Comment (p0lyn0mial) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:83]: > Here's a different idea: make it not depend on RV at all and always work as described here.

actually, this is similar to what I had initially, the only drawback was on resumption (expiration) we would have to wait for etcd vs getting everything from the cache. On the other hand most of the time the watch will get updates. Reinitialization seems to be a rare event. So I am fine with this idea as well.

Comment (lavalamp) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:83]: > when the previous watch finishes with "resource version to old" (or other error that requires relist), what we do is basically we send "List with NotOlderThan semantic with the RV of the last event that we received". 

If one replica is giving a "too old" error for a given RV, that means it's out of the window and it should be impossible for a relist without any RV to get an RV *older* than that, no matter what?

If we always do a consistent read, it should not be possible to go back in time, no?

The only use I can see for accepting an RV, is that if the RV is in the window, we don't need to send initial events at all?

Comment (jpbetz) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:83]: > maybe all we need is to actually don't expose "RV=0" semantic with the new `SendInitialEvents` feature.

Yes. This is the outcome I'd like to see.

> * `RV=0` - we doesn't allow this call and reject those requests via validation
> 
> [And eventually completely deprecate the RV=0 semantic also if `SendInitialEvents` option is not used.]

Yeah, IMHO, almost everywhere RV=0 is called, the caller meant to indicate "I'd like this call to be fast/cached", not "this request can support arbitrarily old data".  And I would really like to not have callers indicate their desire for performance in API requests, so big +1 to deprecation from me.   Any request to RV=0 is also satisfied by RV="" so long as we make it performant enough (presumably serving requests from cache when we can?).

Comment (wojtek-t) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:83]: Sorry for huge delay.

> If one replica is giving a "too old" error for a given RV, that means it's out of the window and it should be impossible for a relist without any RV to get an RV older than that, no matter what?

That's true for etcd, where compaction is a quorum operation. But that's not necessary true for watchcache, where each instance can potentially trim its own history independently from others.
So if one instance is returning too old for a given RV, it doesn't mean the other instance will return that too...


Now, thinking a bit more about that, I think that we focused too much on the `ResourceVersion` aspect. Instead, we should primarily focus on `ResourceVersionMatch` aspect (and treat RV as a parameter for that).

So given we're effectively making a call that is "list+watch" together, the way we should approach is imho is something like that:
```
// SendInitialEvents when set together with Watch option,
// begins the watch stream with synthetic "Added" events of all resources.
// Followed by a synthetic "Bookmark" event containing a ResourceVersion
// after which the server continues streaming events.
//
// When SendInitialEvents option is set, we require  ResourceVersionMatch
// option to also be set. The semantic of the operation is as following:
// - ResourceVersionMatch = NotOlderThan
//        It starts with sending initial events for all objects (at some RV) and
//        potentially followed by an event stream until the state becomes synced
//        to a resource version at least as fresh as the by the ResourceVersion
//        parameter. At this point a synthetic bookmark event is sent and
//        watch stream is continued to be send.
//        If RV is unset, that is interpreted as "consistent read", and watch
//        bookmark will be send when the state will be synced to the moment
//        when the request started being processed.
// - ResourceVersionMatch = Exact
//        NotSupported error is returned
// - ResourceVersionMatch unset (or set to any other version)
//        BadRequest is returned
```

This effectively means that we will still be supporting RV=0, but the semantic is not tight to it's special value, but rather to the concept of "NotOlderThan", which is exactly what we're trying to achieve.

Does that make more sense now?

Comment (wojtek-t) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:83]: I just realized that what we need now, is to effectively open a KEP update and have this discussion on the KEP.
Once the KEP is approved (as an API change), we should update this PR and let it be merged :)

Let me try to open the KEP update for it with the change I proposed above.

Comment (wojtek-t) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/internalversion/types.go:83]: I opened https://github.com/kubernetes/enhancements/pull/3667

I suggest continueing this discussion on that PR (or if needed on some SIG meeting).

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Comment by p0lyn0mial** [comment]
/close

in favour of https://github.com/kubernetes/enhancements/pull/3667
